qiskit_nature/circuit/library/ansaetze/ucc.py

```diff
@@ -0,0 +1,173 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+TODO.
+"""
+
+from functools import partial
+from typing import Callable, List, Optional, Tuple, Union
+
+import logging
+
+from qiskit_nature.components.variational_forms.excitation_op_builder import ExcitationOpBuilder
+from .evolved_operator_ansatz import EvolvedOperatorAnsatz
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: Added to pass mypy, need change to real class
+class QubitOpConverter:
+    """ The QubitOpConverter """
+
+    def convert(self, op):
+        """TODO."""
+        pass
+
+
+class UCC(EvolvedOperatorAnsatz):
+    """The Unitary Coupled-Cluster Ansatz."""
+
+    EXCITATION_TYPE = {
+        's': 1,
+        'd': 2,
+        't': 3,
+        'q': 4,
+    }
+
+    # TODO: expose evolution and reps here, too?
+    def __init__(self, qubit_op_converter: QubitOpConverter,
+                 num_particles: Optional[Tuple[int, int]] = None,
+                 num_spin_orbitals: Optional[int] = None,
+                 excitations: Optional[Union[str, int, List[int], List[Callable]]] = None):
+        """
+
+        Args:
+            qubit_op_converter: the QubitOpConverter instance which takes care of mapping a
+            :code:`~.SecondQuantizedOp` to a :code:`~.PauliSumOp` as well as performing all
+            configured symmetry reductions on it.
+            num_particles: the tuple of the number of alpha- and beta-spin particles.
+            num_spin_orbitals: the number of spin orbitals.
+            excitations: this can be any of the following:
+                - a `str` which contains the types of excitations. Allowed characters are:
+                    - `s` for singles
+                    - `d` for doubles
+                    - `t` for triples
+                    - `q` for quadruples
+                - a single, positive `int` which denotes the number of excitations (1 == `s`, etc.)
+                - a list of positive integers
+                - a list of `callable` objects which are used to generate the excitations. Each
+                  `callable` must take the following arguments:
+                      - `num_particles`: the same as above
+                      - `num_spin_orbitals`: the same as above
+                  and must return a `List[SecondQuantizedOp]`.
+        """
+        self._qubit_op_converter = qubit_op_converter
+        self._num_particles = num_particles
+        self._num_spin_orbitals = num_spin_orbitals
+        self._excitations = excitations
+        # TODO: Added to pass lint, need change
+        super().__init__([], reps=0, evolution=None)
+
+    @property
+    def num_spin_orbitals(self) -> int:
+        """The number of spin orbitals."""
+        return self._num_spin_orbitals
+
+    @num_spin_orbitals.setter
+    def num_spin_orbitals(self, n: int) -> None:
+        """Sets the number of spin orbitals."""
+        self._num_spin_orbitals = n
```

Comment:
If changing the `num_spin_orbitals` changes the circuit structure, the setter must call `self._invalidate` to tell the circuit to rebuild itself 🙂